src/sequence_state.h

```diff
@@ -65,6 +67,17 @@ class SequenceState {
   // The data for this shape.
   std::shared_ptr<Memory>& Data() { return data_; }
 
+  // A boolean indicating whether the state buffer should be reused or not.
+  bool ReuseBuffer() { return reuse_buffer_; }
```

Comment:
Why growable memory requires reuse buffer to be true?

From model config, reuse buffer refers to `use_single_buffer` which control whether the same buffer is used regardless of whether the state is accessed as input or as output. This looks like an optimization config to me (whether there is data copy to update state for input).

Reply:
The main purpose of the growable buffer is to allow the backend to append data to an existing state buffer.

If we don't require use single buffer the backend would always have to write the full output state in each iteration since the output state buffer of step i is different that output state of step i+1. So we can technically allow something like this but I think it defeats the original purpose of this feature.

Reply:
Would it make more sense to only expose growable buffer configuration and make single buffer internal to growable buffer?

Another question is that I don't see why the user do not want to use single buffer in any case, which always saves a data copy within Triton if I understand it correctly.

Reply:
>Would it make more sense to only expose growable buffer configuration and make single buffer internal to growable buffer?

You mean to not expose single buffer optimization externally at all? I think there are scenarios that you want to use single buffer optimization alone just to reduce memory usage (e.g., CPU memory).

> Another question is that I don't see why the user do not want to use single buffer in any case, which always saves a data copy within Triton if I understand it correctly.

It doesn't save a data copy. It reduces the memory usage. But to your original question, I think the issue with always enabling this is that `TRITONBACKEND_StateUpdate` will no longer have an effect (becomes a no-op) which breaks backward compatibility.

Reply:
@Tabrizian ,

I'm wondering if it should be growable automatically if you are using CUDA memory?

My thought is:

1) Input / Output having the same buffer - that seems definitely like an option that a backend may or may not support. That is it's not clear to me that input and output wouldn't be read and written in parallel under some models / architectures. The other option is we can require that they can't be read and written to at the same time and deprecate stateupdate.

2) Growing memory vs re-allocating and copying is transparent to the functionality - so could always be enabled if supported for a buffer type?

Reply:
1. yes, exactly, that's what I mentioned to you earlier.
2. I think there are still some edge cases that we want this to be an option. For example, the minimum allocation for growable memory is 2MBs (depending on the systems) which could be a very large amount esp if you have a lot of sequences. So, I think the user should only use this option for sequences with large memory.